ZA | 25-SDC-July | Luke Manyamazi | Sprint 3 | Implement Shell Tools Exercises

Luke-Manyamazi opened the pull request:
<!--

You must title your PR like this:

Region | Cohort | FirstName LastName | Sprint | Assignment Title

For example,

London | 25-ITP-May | Carol Owen | Sprint 1 | Alarm Clock

Complete the task list below this message.
If your PR is rejected, check the task list.

-->

## Learners, PR Template

Self checklist

- [x] I have committed my files one by one, on purpose, and for a reason
- [x] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)
- [x] My changes meet the [requirements](./README.md) of this task

## Changelist

This PR includes implementations of basic Unix-like command-line tools in Node.js:

ls: Lists files in a directory.
cat: Reads and prints the contents of one or more files.
wc: Counts lines, words, and characters in one or more files, with support for flags (-l, -w, -c).

Each command was tested to match expected output behavior for the given use cases.

## Questions

1. Is my handling of default behavior in the wc command (i.e. when no flags are provided) aligned with how the actual Unix wc works?
2. Do you recommend any improvements for structuring output formatting, especially when dealing with multiple files and totals?
3. Are there any edge cases you suggest I test further?
4.

Luke-Manyamazi committed:
added code for ls exercise

Luke-Manyamazi committed:
added code for the cat file exercises

Luke-Manyamazi committed:
added code for the wc exercise

LonMcGregor commented on implement-shell-tools/cat/cat.mjs:
```diff
@@ -0,0 +1,40 @@
+    const lines = content.split('\n');
+
+    for (const line of lines) {
+      const shouldNumber = (numberAllLines && !numNotBlank) || (numNotBlank && line.trim() !== '');
```
You are not numbering non-blank lines when -b is used. Can you see why that is?

LonMcGregor commented on implement-shell-tools/ls/ls.mjs:
```diff
@@ -0,0 +1,24 @@
+    .name("ls")
+    .description("Lists the files in a directory")
+    .option("-1, --one", "One per line")
+    .option("-a", "Include files starting with dot")
```
You offer -a as an option, but is this ever used?

LonMcGregor commented on implement-shell-tools/wc/wc.mjs:
Throughout this file there is quite a lot of duplication. Can you think of how you might reduce this?

LonMcGregor reviewed (commented):
Good work on this, I've left some comments for where you could start to improve more.

For your questions, yes, it looks like you're handling defaults for wc fine. I think you are handling output fine.

If you are thinking of handling more edge cases, think about how people might use these commands - what sort of arguments might get passed to your code?

Luke-Manyamazi committed:
Refactor shell tools: cat, ls, wc

- cat.mjs:
  - Fixed -b option to number only non-blank lines.
  - Unified line number padding to avoid repetition.
  - Added error handling for directories and missing files.

- ls.mjs:
  - Correctly implemented -a option to include hidden files.
  - Validates single directory argument.
  - Sorted output and handled one-column listing.

- wc.mjs:
  - Centralized line, word, and character counting in countFile().
  - Defaults to counting all metrics if no options are given.
  - Added support for multiple files with totals.
  - Improved error handling for missing files and directories.

Overall:
- Removed duplication, clarified logic, and improved alignment with standard Unix behavior.

LonMcGregor reviewed (commented):
Thanks for making these changes. cat and ls look good now. I still notice some duplication in the wc file however.

Have a look at the last 2 blocks of code. They're both doing the same thing: Check options, push to an array,add padding, then print out. Is there a way you could abstract that code a bit?

Luke-Manyamazi committed:
Refactor wc output formatting  - Abstract duplicated code for printing counts into `formatCounts()` helper - Precompute active count options in `activeCounts` to remove repeated ternaries - Simplify per-file and total output logic for readability and maintainability

LonMcGregor reviewed (commented):
Looks great! You're finished with this task now
